### PR TITLE
Add `Jsonifiable` type

### LINKED ISSUE #356: Proposal: Jsonifyable
I'm hitting a problem where `(string | undefined)[]` is incompatible with `JsonValue`. 

```ts
export type BrokenWrapper<TValue extends JsonValue> = {};

export type MyType = string[];

export type BrokenWrappedType = BrokenWrapper<PartialDeep<MyType>>;
```
https://codesandbox.io/s/nice-water-lp2n5

Will produce:

```
Type '(string | undefined)[]' does not satisfy the constraint 'JsonValue'.
  Type '(string | undefined)[]' is not assignable to type 'JsonArray'.
    Type 'string | undefined' is not assignable to type 'JsonValue'.
      Type 'undefined' is not assignable to type 'JsonValue'.  ts(2344)
```

I think this error is correct because `undefined` is certainly not allowed in a JSON value.

What I want is a way to express an object that *can* pass JSON.stringify without error. Ie, Jsonifyable!

It's almost identical, but `undefined` is allowed as a value:

```ts
export type JsonifyableObject = { [Key in string]?: JsonifyableValue };
export type JsonifyableArray = JsonifyableValue[];
export type JsonifyablePrimitive = string | number | boolean | null | undefined;
//                                                                  ^^^ this is the magic
export type JsonifyableValue =
  | JsonifyablePrimitive
  | JsonifyableObject
  | JsonifyableArray;
```

If you'd like me to turn this into a PR, I'd be happy to. I'm using it in my production code right now.

### PR DESCRIPTION
**Pull request by itsjohncs**
Closes #356.

### CODE REVIEW

**Comment by itsjohncs** [comment]
Whoops, I somehow missed that `undefined` being allowed is desired! Admittedly I did a pretty cursory reading of #356. I'll update the type to permit `undefined`.

File: source/jsonifiable.d.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import type {JsonPrimitive, JsonValue} from './basic';
+
+type JsonifiableObject = {[Key in string]?: Jsonifiable} | {toJSON: () => JsonValue};
+type JsonifiableArray = Jsonifiable[];
+
+/**
+Matches a value that can be losslessly converted to JSON.
+
+`undefined` is allowed in object fields (ex: `{a?: number}`) as a
+special case even though `JSON.stringify({a: undefined})` is `{}`.
```
Comment (sindresorhus) [inline source/jsonifiable.d.ts:10]: Don't hard-wrap.

It think the user will ask themselves why `undefined` is allowed, so I think the docs here should quickly answer that.

File: test-d/jsonifiable.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
+import {expectAssignable, expectNotAssignable} from 'tsd';
+import type {Jsonifiable} from '..';
+
+expectAssignable<Jsonifiable>(1);
+expectAssignable<Jsonifiable>('');
+expectAssignable<Jsonifiable>(null);
+expectAssignable<Jsonifiable>(new Date());
+expectAssignable<Jsonifiable>({a: new Date()});
+expectAssignable<Jsonifiable>([new Date()]);
```
Comment (sindresorhus) [inline test-d/jsonifiable.ts:9]: Can you add an assertion for readonly array too?

File: source/jsonifiable.d.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
+import type {JsonPrimitive, JsonValue} from './basic';
+
+type JsonifiableObject = {[Key in string]?: Jsonifiable} | {toJSON: () => JsonValue};
+type JsonifiableArray = Jsonifiable[];
+
+/**
+Matches a value that can be losslessly converted to JSON.
+
```
Comment (sindresorhus) [inline source/jsonifiable.d.ts:8]: Would you be useful to mention some use-cases for this type. (I realize there is an example, but nothing beats just reading some text quickly)

File: source/jsonifiable.d.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
+import type {JsonPrimitive, JsonValue} from './basic';
+
+type JsonifiableObject = {[Key in string]?: Jsonifiable} | {toJSON: () => JsonValue};
```
Comment (sindresorhus) [inline source/jsonifiable.d.ts:3]: Shouldn't this look deeply for `toJSON`?

See https://github.com/sindresorhus/type-fest/commit/e11d30b727fc8b3e9438fbf15ec8619449e5566e and https://github.com/sindresorhus/type-fest/issues/468

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: source/jsonifiable.d.ts (hunk 4 above)
Comment (itsjohncs) [reply source/jsonifiable.d.ts:3]: Could you give me a test case that you're worried about? I'm not sure I understand your concern. I added this test case as a guess of what you mean though:

```typescript
expectAssignable<Jsonifiable>({a: {deeply: {nested: {toJsonObject: new Date()}}}});
```

Comment (itsjohncs) [reply source/jsonifiable.d.ts:3]: Disregard my last comment. I must have messed up the test I did which prompted https://github.com/sindresorhus/type-fest/pull/492/commits/d8d78f8c052561b8fefbe431c95cb6f507e2b8e1. I'll roll back that change and add a test.

**Review by sindresorhus — APPROVED** [review]

**Review by skarab42 — APPROVED** [review]

**Comment by MichaelDeBoey** [comment]
Should we use this type in `Jsonify` as well?

**Comment by skarab42** [comment]
> Should we use this type in `Jsonify` as well?

No
- `Jsonifiable` is used to type values that you expect to pass to `JSON.stringify` where { foo: undefined } and plain {} are functionally equivalent.
- `Jsonify` is used to type parsed JSON value, coming from eg. JSON.parse() where undefined as a value is impossible.

FI https://github.com/sindresorhus/type-fest/issues/471#issuecomment-1285314479

**Comment by shaharmor** [comment]
@skarab42 why no also export `JsonifiableObject` and `JsonifiableArray`?

**Pull request merged**